### PR TITLE
51145 & 56174

### PR DESCRIPTION
**Pull request by jane-intel**
### Details:
 - *propagate output shape dynamically*

### Tickets:
 - *51145*

### CODE REVIEW

File: src/common/transformations/include/transformations/dimension_tracking.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#pragma once
+
+#include <memory>
+
+#include <transformations_visibility.hpp>
+
+#include <openvino/pass/graph_rewrite.hpp>
+#include <openvino/opsets/opset1.hpp>
+
+using namespace std;
```
Comment (nosovmik) [inline src/common/transformations/include/transformations/dimension_tracking.hpp:14]: suggesting remove this, it is always recommended avoid 'using namespace' in header files

File: src/common/transformations/include/transformations/dimension_tracking.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+using P2Btype = std::unordered_map<ngraph::op::Parameter*, std::unordered_set<size_t>>;
+
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API FindBatch;
```
Comment (nosovmik) [inline src/common/transformations/include/transformations/dimension_tracking.hpp:20]: can this forward declaration be removed? FindBatch definition follows right after this declaration

File: src/common/transformations/include/transformations/dimension_tracking.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+class DimensionTracker;
+
+namespace batch_util {
+    std::map<ov::opset1::Parameter*, ov::PartialShape> collect_original_shapes(std::shared_ptr<ov::Model> f);
+    void mark_batch(ov::opset1::Parameter *parameter, P2Btype &map, const std::unordered_set<size_t> &batches);
+    void mark_no_batch(ov::opset1::Parameter *parameter, P2Btype &map, ov::Node *node);
+    void mark_layout_independent_batch(ov::opset1::Parameter *parameter, ov::Node *result, P2Btype &map);
+    void mark_with_unique_dimension_labels(const std::shared_ptr<Model> &f, const ov::DimensionTracker &dt);
+    void restore_original_dimensions_except_batch(const std::map<ov::opset1::Parameter*, ngraph::PartialShape>& parameter_to_shape);
+    bool check_batch_tracks_through_all_the_nodes(std::shared_ptr<ov::Model> f);
```
Comment (nosovmik) [inline src/common/transformations/include/transformations/dimension_tracking.hpp:41]:
```suggestion
    bool check_batch_tracks_through_all_the_nodes(const std::shared_ptr<ov::Model>& m);
```

**Review by nosovmik — COMMENTED** [review]
Please rebase

**Comment by jane-intel** [comment]
@nosovmik sorry -- it was a bad merge. Thank you for suggestions -- they are all addressed on master already.

**Review by rnugmanx — APPROVED** [review]

**Pull request merged**